### PR TITLE
Avoid boxing variables that may be used undef

### PR DESCRIPTION
**Pull request by carnaval**
use a separate `i1` flag so that we avoid boxing and give visibility to llvm in the defined/undefined dataflow, to remove redundant branches etc

``` julia
f(a,b) = (if a; c = b; end; c)
```

gives

```
define i64 @julia_f_65179(i8, i64) #0 {
top:
  %2 = and i8 %0, 1
  %3 = icmp eq i8 %2, 0
  br i1 %3, label %err, label %ok

err:                                              ; preds = %top
  call void @jl_undefined_var_error(%jl_value_t* inttoptr (i64 140728621134648 to %jl_value_t*))
  unreachable

ok:                                               ; preds = %top
  ret i64 %1
}
```

### CODE REVIEW

File: src/codegen.cpp [hunk 1]
```diff
@@ -490,6 +490,9 @@ struct jl_varinfo_t {
 #else
     DIVariable dinfo;
 #endif
+    // if the variable might be used undefined and is not boxed
+    // this i1 flag is true when it is defined
+    Value *undefFlag;
```
Comment (JeffBezanson) [inline src/codegen.cpp]: I suggest calling it `defFlag` to avoid double negatives :)

**Review by JeffBezanson — APPROVED** [review]

**Comment by JeffBezanson** [comment]
Awesome; this has been a long time coming.

File: src/codegen.cpp [hunk 2]
```diff
@@ -737,6 +741,21 @@ static jl_sym_t *slot_symbol(int s, jl_codectx_t *ctx)
...
+{
+    assert(!vi.memloc); // undefinedness is null pointer for boxed things
+    assert(vi.usedUndef && vi.undefFlag);
+    builder.CreateStore(ConstantInt::get(T_int1, val), vi.undefFlag);
+}
+
+static void alloc_undef_flag(jl_varinfo_t& vi, jl_codectx_t* ctx)
+{
+    if (vi.usedUndef) {
+        vi.undefFlag = emit_static_alloca(T_int1, ctx);
```
Comment (yuyichao) [inline src/codegen.cpp]: We only need this if the slot is a `isbits` type right?

Comment (carnaval) [inline src/codegen.cpp]: or a constant. It should only be called in those cases. I'll add an assertion.

File: src/codegen.cpp (hunk 1 above)
Comment (carnaval) [inline src/codegen.cpp]: yep, that was gratuitously confusing

**Comment by tkelman** [comment]
we need to remember to do this more often: @nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/18732#issuecomment-250631988) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/90d960f1e46fedaa284d8bdbb381a710b4ee1b08/8ff9156_vs_b36141f/report.md). cc @jrevels

**Review by StefanKarpinski — CHANGES_REQUESTED** [review]

**Comment by JeffBezanson** [comment]
Bump. Just needs a bit more renaming, then good to go.

**Comment by JeffBezanson** [comment]
Ok, did it myself :P

**Comment by JeffBezanson** [comment]
Fixes #6914

**Comment by vtjnash** [comment]
I think there was a bug in this.
```
Assertion failed: (vi.defFlag), function emit_local, file /Volumes/Lion/Users/jameson/Documents/julia/src/codegen.cpp, line 3142.

signal (4): Illegal instruction: 4
while loading /Volumes/Lion/Users/jameson/Documents/julia/base/precompile.jl, in expression starting on line 480
abort at /Volumes/Lion/Users/jameson/Documents/julia/build-llvm37/usr/lib//libLLVM-3.7.dylib (unknown line)
__assert_rtn at /Volumes/Lion/Users/jameson/Documents/julia/build-llvm37/usr/lib//libLLVM-3.7.dylib (unknown line)
emit_local at /Volumes/Lion/Users/jameson/Documents/julia/src/codegen.cpp:3142 [inlined]
emit_expr at /Volumes/Lion/Users/jameson/Documents/julia/src/codegen.cpp:3338
emit_assignment at /Volumes/Lion/Users/jameson/Documents/julia/src/codegen.cpp:3217 [inlined]
emit_expr at /Volumes/Lion/Users/jameson/Documents/julia/src/codegen.cpp:3421
emit_stmtpos at /Volumes/Lion/Users/jameson/Documents/julia/src/codegen.cpp:3327 [inlined]
emit_function at /Volumes/Lion/Users/jameson/Documents/julia/src/codegen.cpp:5081
jl_compile_linfo at /Volumes/Lion/Users/jameson/Documents/julia/src/codegen.cpp:927
jl_generate_fptr at /Volumes/Lion/Users/jameson/Documents/julia/src/codegen.cpp:1170
```
```julia
CodeInfo(code=Array{Any, 1}[
  # line 312,
  Expr(:meta, :push_loc, :multidimensional.jl)::Any,
  # line 315,
  SlotNumber(id=10) = SlotNumber(id=3),
  SlotNumber(id=8) = SlotNumber(id=1),
  SlotNumber(id=9) = SlotNumber(id=2),
  SlotNumber(id=11) = SlotNumber(id=4),
  SlotNumber(id=5) = Expr(:call, Base.to_index, Expr(:call, Base.getindex, SlotNumber(id=11), 1)::Any)::Any,
  # line 316,
  SlotNumber(id=6) = Expr(:call, Base.index_shape, SlotNumber(id=10), SlotNumber(id=5))::Any,
  # line 317,
  SlotNumber(id=7) = Expr(:call, Base.similar, SlotNumber(id=10), SlotNumber(id=6))::Any,
  # line 318,
  SSAValue(0) = Expr(:call, Base.==, Expr(:call, Base.map, Base.unsafe_length, Expr(:call, Base.indices, SlotNumber(id=7))::Any)::Any, Expr(:call, Base.map, Base.unsafe_length, SlotNumber(id=6))::Any)::Any,
  Expr(:gotoifnot, SSAValue(0), 17)::Any,
  goto 19,
  17:,
  Expr(:call, Base.throw_checksize_error, SlotNumber(id=7), SlotNumber(id=6))::Any,
  19:,
  # line 319,
  SSAValue(1) = Expr(:call, Base._unsafe_getindex!, SlotNumber(id=7), SlotNumber(id=10), SlotNumber(id=5))::Any,
  Expr(:meta, :pop_loc)::Any,
  Expr(:return, SSAValue(1))::Any], slottypes=nothing, ssavaluetypes=2, slotnames=Array{Any, 1}[
  :#self#,
  :#unused#,
  :A,
  :I,
  :I_1,
  :shape,
  :dest,
  :#self#,
  :#unused#,
  :A,
  :I], slotflags=Array{UInt8, 1}[0x02, 0x02, 0x02, 0x02, 0x12, 0x12, 0x12, 0x02, 0x02, 0x02, 0x02], inferred=false, inlineable=false, propagate_inbounds=false, pure=false)
```

**Comment by JeffBezanson** [comment]
I have a fix.

**Comment by JeffBezanson** [comment]
#19036

**Pull request merged**